### ISSUE TITLE
Flow Typecast Comments

### ISSUE THREAD

**Issue by jacobp100** [prettier/prettier#719]
In flow, you can typecast with the following,

```js
(obj: Class).property
```

You can also do this with comments,

```js
(obj /*: Class */).property
```

However, this gets transformed into,

```js
obj /*: Class */.property
```

Both are syntactically valid, but only one is recognised by flow. Maybe we'd just want to check comments that start with `:`?

Happy to help if you decide what direction we should go!

# Edit

Also,

```js
const x = (input /*: string */) /*: string */ => {};
// to
const x = (input /*: string */ /*: string */) => {};
```

**Comment by vjeux** [comment]
Thanks for the report, we need to fix it!

**Comment by vjeux** [comment]
@jacobp100 https://github.com/jlongster/prettier/pull/722

Do you mind running this pull request and try to find crazy annotations that wouldn't work with it? Thanks!

**Comment by jacobp100** [comment]
Seemed to work!

**Comment by vjeux** [comment]
Another idea: use the flow parser to parse comments and have a way to output flow types as comments instead of the ast version.

**Comment by jokeyrhyme** [comment]
Related to #204 ?

**Comment by jokeyrhyme** [comment]
I've just encountered the Flow comment syntax issues with functions

Here are a few more test cases, I hope these are helpful

https://prettier.github.io v0.22.0

```js
// input
function fnNoArgs () /* : Promise<void> */ {
  return Promise.resolve()
}

// output
function fnNoArgs(/* : Promise<void> */) {
  return Promise.resolve();
}
```

```js
// input
function fnOneArg (a /* : string */) /* : Promise<string> */ {
  return Promise.resolve(a)
}

// output
function fnOneArg(a /* : string */) /* : Promise<string> */ {
  return Promise.resolve(a);
}
```

```js
// input
const arrowNoArgs = () /* : Promise<void> */ => {
  return Promise.resolve()
}

// output
const arrowNoArgs = (/* : Promise<void> */) => {
  return Promise.resolve();
};
```

```js
// input
const arrowOneArge = (a /* : string */) /* : Promise<string> */ => {
  return Promise.resolve(a)
}

// output
const arrowOneArge = (a /* : string */ /* : Promise<string> */) => {
  return Promise.resolve(a);
};
```

**Comment by jokeyrhyme** [comment]
Awesome work! It looks like most of the above has been resolved.

As of 1.2.2, the first two examples [above](https://github.com/prettier/prettier/issues/719#issuecomment-286396699) are now working perfectly. That is, these are the cases with traditional `function`s with Flow comments.

Arrow functions (`=>`) still seem to be problematic, in that the return type Flow comment is being positioned in the wrong place:

```js
// input
const arrowNoArgs = () /* : Promise<void> */ => {
  return Promise.resolve()
}

```

```js
// input
const arrowOneArge = (a /* : string */) /* : Promise<string> */ => {
  return Promise.resolve(a)
}

// output
const arrowOneArge = (a /* : string */ /* : Promise<string> */) => {
  return Promise.resolve(a);
};
```

**Comment by vjeux** [comment]
We now have a much better infrastructure for dealing with comments so we should be able to fix all the issues reported here :)

**Comment by brownieboy** [comment]
Any update on this one?

Comments are still being moved from outside to inside on arrow functions for me.  I'm using Prettier 1.7.0.

**Comment by duailibe** [comment]
All the examples off arrow and regular functions are working correctly in master now. You can check out this preview (I copied all the samples in this issue): [link](https://deploy-preview-3444--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEAKCAjAVgAgPQBUSOAwgDYCGAzlTgXgJQB0ADgE4QtxswCeA3AB0owyFCowcADxwBeHKgCWUFgFdJhYhLbKA5nUb4iObXoNyAfDmABfISKgAzVVDAxF0HI6gA5CAEE2XVpUBiMcYgAFDgBbRSo4AB4ANwhFABMremthHBw2OBhVNigcaIg4hKYCqggyZLhQ4RthYWdXd09vAHkoOED9VApwrRgdKH16MMIIstj4pNMJrLwc0vzC4tLyyrhquFr6xooGZtbXaAkcCjYOAHc-Adp5UJG5ioWUtMzzWStgXIbIold67faHBpNKAtBxiK43e69fpBOByBTDGajcaTQyY0GfJa6FaWNZ5ArA7bzKo1OqQk5nKAgAA0IE4nXEyFACIgd0iNwQVGQIAoqQyzOFEmQjgoZASLIwbAoYAA1oUAMosJV6ZBjVRwFnpCBgKUyuUgZQJHjRCi6GIUE2y-UgLBUKQAIUVKvVFBicAAMso4A6zRB1GoYAAmYNOzVsS1CjAUDC8MjQcXsZQwADqGRgAAtkAAOAAMLPYEASWcVLCF7AO3Aa4oKAEdVIoCtbbfakNLHSyEnEdWw9f29GQ4ABFVQQeDRlkwJM59L55AR+eKxRkPQkCp2oVQaBBlmqBIAFSTgp7prgNhsQA)

Will work on the case OP mentioned

**Comment by akalin** [comment]
Trying 1.10, this fixes a lot of cases, thanks! One remaining case I ran into is:

```
    this.state = ({ a, b } /* : AddDemoState */);
```

[link](https://deploy-preview-3444--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEMAWBLAzgOizAQ3gAIBeYgCmGIIBpiAjYgX2IHoAqYpYgQQBN+AETgBbCAGVCJDmwCUAbhC0QEAA4wM0LMlAEATvogB3AAoGEOlAQBuEDP2UgC+ZADMCAGyxwVDfQRgANZwMBJqgRhQAObIMPoArr4g-BBg7l4+KlE++jCmAdGiBBneyQBWWAAeAEIBwaESBKJwADJRcKVZqgkwar0ATF3JEfq5yCAMBAwAnp7QTmr6UTAA6g7oyAAcAAwqSxA+qwFqE0twuTadKvpwAI4JGLcFBEUlSB5lKj6iGHGJySwUWinjgAEUEhB4MMVIQGOt+JskANYQEMJ5gQBhCCiYoTKDQa4gBI+AAq0ysnx8zGYQA)

**Comment by suchipi** [comment]
As a workaround for that particular case, I think you could do:
```
this.state /* : AddDemoState */ = { a, b };
```

**Comment by akalin** [comment]
Ah, just tried that. flow doesn't seem to like that, I'm guessing because `state` is a field and not a variable.

**Comment by suchipi** [comment]
I guess you could do this as a workaround then:
```js
const addDemoState /* : AddDemoState */ = { a, b };
this.state = addDemoState;
```

**Comment by suchipi** [comment]
(We still want to fix this long-term though)

**Comment by k7n4n5t3w4rt** [comment]
I'm not sure I'm understanding this correctly but I'm having problems with flow types and generics using the comment syntax.

This:

```
/*::
type Props = {
  prop1: string,
  prop2: number
}
*/
class App extends Component/*:: <Props> */ {

...

}
```

is becoming this:

```
type Props = {
  prop1: string,
  prop2: number
}
class App extends Component<Props> {

...

}
```

which is valid flow syntax of course. However, I was expecting the comment syntax to be preserved. Am I missing something?

**Comment by hawkrives** [comment]
@k7n4n5t3w4rt Can you reproduce that bug on the playground (prettier.io)? I ask because I've never personally experienced Prettier changing my type comments to code.

**Comment by k7n4n5t3w4rt** [comment]
@hawkrives It seems to be doing the same thing in the playground when the parser option is set to "flow":

https://goo.gl/pL1Nwu

**Comment by duailibe** [comment]
@k7n4n5t3w4rt It's a known bug, you can follow it on #3493

**Comment by hawkrives** [comment]
@k7n4n5t3w4rt Ah, you're using the Flow parser. 

If you don't _need_ to use it, you can use the (default) Babylon parser, which also supports Flow types, and doesn't have this bug: [playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlKagFRJKkwCeADnBgCpwDOMGAvBsKcRgMAThCbD6ARiQZOwgJZQA5gBpSGDRhFi4EugCYZUAK4BbAEa7SAX1IVUAqCBUgxMedHbJQAQ2GjCAAU-BC8UHwA3CHkAE2cQc2EfMABrOBgAZQZkxSVkGGFjOBdFdl0YQKSlUx9kHB88MpcAK3YADwAhJNT0jJ9TOAAZRTg6hqaQbOEy4WQEn3M6AicXEUUYAHVYmAALZAAOAAZV0TKNpIY5kQ5dCNGXYTgAR2N5R8qfatqkesbikDKpnk+UK-3YuTwcAAisYIPAxn8XDAFlsYrtkPokUl5HhcgBhCCmGpzKDQe4gYxlFgLMK-MrWaxAA)

**Comment by k7n4n5t3w4rt** [comment]
@hawkrives Aha! I didn't realise I could use the babylon parser, thanks.

@duailibe Thanks for that, I'll keep an eye on it.

**Comment by pannous** [comment]
can prettier be used to convert flow annotations into flow comments?
like thus:
```
prettier --emit-flow-comments   "var test:int=8"
var test/*:int*/=8
```

**Comment by duailibe** [comment]
@pannous I don't think so, it's out of Prettier's scope, but it's really interesting and someone should totally build a tool that does that.

**Comment by suchipi** [comment]
@pannous you can use https://www.npmjs.com/package/@babel/plugin-transform-flow-comments to do that, though the formatting isn't great:

```js
const { code } = require("@babel/core").transform("var test:int=8", {
  plugins: [require("@babel/plugin-transform-flow-comments")]
});
console.log(code); // "var test\n/*:int*/\n= 8;"
```

**Comment by swac** [comment]
What's the overall status of this? It seems there are two options, but neither is actually compatible with using Flow's comment syntax.

The first option is to specify `"flow"` as the parser in Prettier's config file. This preserves Flow compatibility, but it replaces Flow comments with Flow's standard syntax, as shown in @k7n4n5t3w4rt's [comment](https://github.com/prettier/prettier/issues/719#issuecomment-390181079). This effectively defeats the purpose of Flow's comment syntax.

The second option is to stick with the default parser, which retains Flow's comments, but removes the necessary parentheses to satisfy Flow. So something like this:

```
new (window.Request/*: Class<Request> */)('');
```

[gets turned into this:](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlKVHIRgBSECWUAJkQHQBKcAjgK5wDOMVACokGAMJ4Ahnz4AeDj34wAfBiGoAlNQDk2jQG4QAGhAQADjDrQ+yUJIBO9ogAUHCGykkA3CHUbGQACN7STAAazgYAGUzUIYAc2QYe14TBj44exhnEPiAW0lkHEk8DJMAKz4ADwAhEPDIqMk8uAAZBjgikrKQWPsM+2QgyUCATwIoALN7BhgAdT8YAAtkAA4ABhNpiAy5kLMh6f5Mr06Tey5uOgucyXzCpGLSuBMMvLoklJeQPgS8OAAitwIPAus8TDARgtGMtkAAmCEhOh4BJiCB5ApDKDQM4gbgZAAqIw8TwyAF8yUA)

```
new window.Request /*: Class<Request> */("");
```

This is no longer valid Flow syntax, so Flow throws an error.

Am I correct that because of these two issues together, there's effectively no way to use Flow's comment syntax with Prettier (barring heavy use of `prettier-ignore` comments)? Although [it's possible to rewrite the code](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlKZFAM4wYBKcAjgK5zWoBUSGAwngIaVKAHnrNWMAHwZ2mALwZCASygATIgDpRLagG5SUOITqNtMABQByCwEodIADQgIABxiLolZKD4AnH0QAFXwRPFD4ANwhFFQcQACMfPjAAazgYAGVnJOUAc2QYHxZHZUo4HxgAxJyAWz5kHD48UscAK0oADwAhRJS09L5quAAZZTh6xuaQLJ9Sn2R4vjiATwIoWOcfZRgAdWiYAAtkAA4ABkcNiFLtxOd5jdYy8LHHHxNFV8q+GrqkBqa4RylaqKfKFAEgSi5PBwACKTAg8HG-0cMEWuxUB2QACYUYlFHhctwINVavMoNBniAmKUACqLUJ-UoAXyZQA) like:

```
const Request/*: Class<Request> */ = window.Request;
new Request('');
```

this slightly defeats the purpose of Prettier, since it no longer serves as an autoformatter that you can just forget about.

**Comment by evilebottnawi** [comment]
@swac PR welcome

**Comment by swac** [comment]
@evilebottnawi Apologies if my comment sounded overly critical. I was just trying to sum up the current state of affairs, since it took me a while to piece together everything from this comment thread 😃

**Comment by evilebottnawi** [comment]
@swac don't worry, we have limited time and contributor to fix all problems, do it yourself it's true the quickest solution :smile:

**Comment by swac** [comment]
@evilebottnawi I took a stab at fixing this in https://github.com/prettier/prettier/pull/5280 😄

**Comment by ikatyang** [comment]
Fixed by #5280.
